channeldb/db.go

```diff
@@ -263,3 +296,123 @@ func (d *DB) FetchAllChannels() ([]*OpenChannel, error) {
 
 	return channels, err
 }
+
+// SyncVersions function is used for safe db version synchronization. It applies
+// migration functions to the current database and recovers the previous
+// state of db if at least one error/panic appeared during migration.
+func (d *DB) SyncVersions(versions []Version) (returnErr error) {
+	meta, err := d.FetchMeta()
+	if err != nil {
+		return err
+	}
+
+	cleanup, err := d.createBackup()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		if r := recover(); r != nil {
+			returnErr = fmt.Errorf("panic occured: %v",
+				errors.New(r))
+		}
+
+		if returnErr != nil {
+			if err := cleanup(true); err != nil {
+				returnErr = fmt.Errorf("while restoring db "+
+					"because of error: '%v', another "+
+					"error is occured: '%v'", returnErr, err)
+			}
+		} else {
+			if err := cleanup(false); err != nil {
+				returnErr = fmt.Errorf("migration was proceeded "+
+					"without problems, but while backup "+
+					"cleanup error is occured: %v", err)
+			}
+		}
+	}()
+
+	latestVersion := getLatestDBVersion(versions)
+
+	if meta.dbVersionNumber < latestVersion {
+		migrations := getMigrationsToApply(versions, meta.dbVersionNumber)
+
+		for _, migration := range migrations {
+			if migration == nil {
+				continue
+			}
+
+			if err := migration(d); err != nil {
+				returnErr = err
+				return
+			}
+		}
+
+		meta.dbVersionNumber = latestVersion
+		if err := d.PutMeta(meta); err != nil {
+			returnErr = err
+			return
+		}
+	}
+
+	return
+}
+
+func getLatestDBVersion(versions []Version) uint32 {
+	return versions[len(versions)-1].number
+}
+
+// getMigrationsToApply retrieves the migration function that should be
+// applyied to the database.
+func getMigrationsToApply(versions []Version, version uint32) []Migration {
+	migrations := make([]Migration, 0)
```

Comment:
Just a minor nit, but alternative syntax which achieves the same thing is: 
```go
var migrations []Migration
```

Alternatively, we could also do something like this which only results in a single allocation: 
```go
migrations := make([]Migration, 0, len(versions))
```